src/components/Canvas.tsx

```diff
@@ -1,70 +1,104 @@
 import * as React from "react";
 
-import {Circle} from "./Circle.tsx";
-import {Square} from "./Square.tsx";
-import {Triangle} from "./Triangle.tsx";
+import {CanvasItem} from "./CanvasItem.tsx";
 
 import * as models from "../models.ts";
 
 export interface CanvasProps {
 }
 
 export interface CanvasState {
-  shapes: models.Shape[];
+  shapes?: models.Shape[];
+  draggingShape?: models.Shape;
+  draggingShapeOffsetX?: number;
+  draggingShapeOffsetY?: number;
 }
 
 export class Canvas extends React.Component<CanvasProps, CanvasState> {
   constructor(props: CanvasProps) {
     super(props);
-    this.state = {shapes: []};
+    this.state = {
+      shapes: [],
+      draggingShapeOffsetX: 0,
+      draggingShapeOffsetY: 0
+    };
   }
 
   render() {
-    return <div className="canvas" onDrop={(e) => {this.drop(e)}} onDragOver={(e) => {this.dragOver(e)}}> 
+    return <div className="canvas"
+      onDragOver={(e) => {this.dragOver(e)}}
+      onDrop={(e) => {this.drop(e)}}
+      onMouseMove={(e) => {this.mouseMove(e)}}
```

Comment:
Instead of adding `onMouseMove` and `onMouseUp` to these elements, I'd suggest adding them to window when `onMouseDown` gets called. This way you'll get move events even when the user drags their mouse off the canvas, and you'll get a mouse up event even if the user drags the element fast enough that the mouse is no longer over the dragging element.